fluture.js

```diff
@@ -376,544 +404,927 @@
+      return this._f(rej, res);
+    },
+
+    inspect: function Future$inspect(){
+      return this.toString();
+    },
+
+    value: function Future$value(f){
+      check$value(this, f);
+      return this._f(
+        function Future$value$rej(e){
+          throw new Error(
+            `Future#value was called on a rejected Future\n  Actual: Future.reject(${show(e)})`
+          );
+        },
+        f
+      );
+    },
+
+    promise: function Future$promise(){
+      check$promise(this);
+      const _this = this;
+      return new Promise(function Future$promise$do(resolve, reject){
+        _this._f(reject, resolve);
+      });
+    }
+
+  };
+
+  //Static functions.
+  Future.of = Future$of;
+  Future.ap = createUnaryDispatcher(FL.ap);
+  Future.map = createUnaryDispatcher(FL.map);
+  Future.bimap = createBinaryDispatcher(FL.bimap);
+  Future.chain = createUnaryDispatcher(FL.chain);
+  Future.chainRec = Future$chainRec;
+  Future.recur = createUnaryDispatcher('recur');
+  Future.chainRej = createUnaryDispatcher('chainRej');
+  Future.mapRej = createUnaryDispatcher('mapRej');
+  Future.swap = createNullaryDispatcher('swap');
+  Future.fork = createBinaryDispatcher('fork');
+  Future.race = createUnaryDispatcher('race');
+  Future.or = createUnaryDispatcher('or');
+  Future.fold = createBinaryDispatcher('fold');
+  Future.hook = createInvertedBinaryDispatcher('hook');
+  Future.finally = createUnaryDispatcher('finally');
+  Future.value = createUnaryDispatcher('value');
+  Future.promise = createNullaryDispatcher('promise');
+  Future.cache = createNullaryDispatcher('cache');
+  Future.Future = Future;
+  Future.isFuture = isFuture;
+  Future.isForkable = isForkable;
+
+  Future.reject = function Future$reject(x){
+    return new FutureReject(x);
+  };
+
+  Future.after = function Future$after(n, x){
+    if(arguments.length === 1) return unaryPartial(Future$after, n);
+    check$after(n);
+    return new FutureAfter(n, x);
+  };
+
+  Future.cast = function Future$cast(m){
+    check$cast(m);
+    return new FutureCast(m);
+  };
+
+  Future.try = function Future$try(f){
+    check$try(f);
+    return new FutureTry(f);
+  };
+
+  Future.encase = function Future$encase(f, x){
+    check$encase(f);
+    if(arguments.length === 1) return unaryPartial(Future.encase, f);
+    return new FutureEncase(f, x);
+  };
+
+  Future.encase2 = function Future$encase2(f, x, y){
+    check$encase2(f);
+    if(arguments.length === 1) return unaryPartial(Future.encase2, f);
+    if(arguments.length === 2) return binaryPartial(Future.encase2, f, x);
+    return new FutureEncase(f, x, y);
+  };
+
+  Future.encase3 = function Future$encase3(f, x, y, z){
+    check$encase3(f);
+    if(arguments.length === 1) return unaryPartial(Future.encase3, f);
+    if(arguments.length === 2) return binaryPartial(Future.encase3, f, x);
+    if(arguments.length === 3) return ternaryPartial(Future.encase3, f, x, y);
+    return new FutureEncase(f, x, y, z);
```

Comment:
Have you considered using `switch` for these?

```javascript
switch (arguments.length) {
  case 1: return unaryPartial(Future.encase3, f);
  case 2: return binaryPartial(Future.encase3, f, x);
  case 3: return ternaryPartial(Future.encase3, f, x, y);
  default: return new FutureEncase(f, x, y, z);
}
```

Also, why do you refer to `Future.encase3` rather than `Future$encase3`? A recursive function is more self-contained if it refers to itself by name rather than via the identifier to which it happens to be bound.